From 7f8cc09deef4b5742d2fa122977844d3812bd318 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Thu, 29 Nov 2012 10:48:07 -0500
Subject: [PATCH] Mob Spawning Events

Add 2 events for better control of mob spawning (Natural and Spawners) This allows us a much more powerful control over spawning and improve performance
---
 src/main/java/net/minecraft/server/EntityHuman.java    |  1 +
 .../java/net/minecraft/server/MobSpawnerAbstract.java  | 18 +++++++++++++++++-
 .../java/net/minecraft/server/SpawnerCreature.java     |  8 ++++++++
 src/main/java/net/minecraft/server/World.java          |  6 ++++++
 4 files changed, 32 insertions(+), 1 deletion(-)

diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index fd08195cf..45439f161 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -64,6 +64,7 @@ public abstract class EntityHuman extends EntityLiving {
     @Nullable
     public EntityFishingHook hookedFish;
     public boolean affectsSpawning = true;
+    static com.google.common.base.Predicate<Entity> affectsSpawningPredicate = (player) -> (player instanceof EntityHuman && ((EntityHuman) player).affectsSpawning); // EMC
 
     // CraftBukkit start
     public boolean fauxSleeping;
diff --git a/src/main/java/net/minecraft/server/MobSpawnerAbstract.java b/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
index e523267e3..51ec23754 100644
--- a/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
+++ b/src/main/java/net/minecraft/server/MobSpawnerAbstract.java
@@ -42,7 +42,23 @@ public abstract class MobSpawnerAbstract {
     private boolean h() {
         BlockPosition blockposition = this.b();
 
-        return this.a().isPlayerNearby((double) blockposition.getX() + 0.5D, (double) blockposition.getY() + 0.5D, (double) blockposition.getZ() + 0.5D, (double) this.requiredPlayerRange);
+        //return this.a().isPlayerNearby((double) blockposition.getX() + 0.5D, (double) blockposition.getY() + 0.5D, (double) blockposition.getZ() + 0.5D, (double) this.requiredPlayerRange); // Paper - affectsSpawning filter
+
+        // EMC start - Add in SpawnerInitiateEvent
+        final World world = this.a();
+        final int x = blockposition.getX();
+        final int y = blockposition.getY();
+        final int z = blockposition.getZ();
+        EntityHuman entity = world.findNearbyPlayer((double) x + 0.5D,
+            (double) y + 0.5D,
+            (double) z + 0.5D,
+            (double) this.requiredPlayerRange, EntityHuman.affectsSpawningPredicate);
+        if (entity == null) {
+            return false;
+        }
+        org.bukkit.Location loc = new org.bukkit.Location(world.getWorld(), x, y, z);
+        return new com.empireminecraft.customevents.SpawnerInitiateEvent(this.getMobName(), world.getWorld(), loc, entity.getBukkitEntity()).callEvent();
+        // EMC end
     }
 
     public void c() {
diff --git a/src/main/java/net/minecraft/server/SpawnerCreature.java b/src/main/java/net/minecraft/server/SpawnerCreature.java
index c1f6e63f6..9d3337327 100644
--- a/src/main/java/net/minecraft/server/SpawnerCreature.java
+++ b/src/main/java/net/minecraft/server/SpawnerCreature.java
@@ -73,6 +73,14 @@ public final class SpawnerCreature {
                     b0 = ( b0 > worldserver.spigotConfig.viewDistance ) ? (byte) worldserver.spigotConfig.viewDistance : b0;
                     b0 = ( b0 > 8 ) ? 8 : b0;
 
+                    // EMC start - SpawnTickEvent
+                    com.empireminecraft.customevents.SpawnTickEvent event = new com.empireminecraft.customevents.SpawnTickEvent(entityhuman.getBukkitEntity(), b0);
+                    if (!event.callEvent()) {
+                        continue;
+                    }
+                    b0 = event.getSpawnRadius();
+                    // EMC end
+
                     for (int i1 = -b0; i1 <= b0; ++i1) {
                         for (k = -b0; k <= b0; ++k) {
                             boolean flag4 = i1 == -b0 || i1 == b0 || k == -b0 || k == b0;
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 02f2bdc44..a1f16eaed 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -2839,6 +2839,12 @@ public abstract class World implements IBlockAccess {
         return this.a(d0, d1, d2, d3, predicate);
     }
 
+    // EMC start // OBF HELPER
+    @Nullable
+    public EntityHuman findNearbyPlayer(double d0, double d1, double d2, double d3, Predicate<Entity> predicate) {
+        return this.a(d0, d1, d2, d3, predicate);
+    }
+    // EMC end
     @Nullable
     public EntityHuman a(double d0, double d1, double d2, double d3, Predicate<Entity> predicate) {
         double d4 = -1.0D;
-- 
2.11.0

